cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/sidecar/ArtifactLocalizer.java

```diff
@@ -0,0 +1,331 @@
+  private final RetryStrategy retryStrategy;
+  private final String basePath;
+
+  @Inject
+  ArtifactLocalizer(CConfiguration cConf, DiscoveryServiceClient discoveryServiceClient) {
+    this(cConf, discoveryServiceClient, "");
+  }
+
+  @VisibleForTesting
+  public ArtifactLocalizer(CConfiguration cConf, DiscoveryServiceClient discoveryServiceClient, String basePath) {
+    this.remoteClient = new RemoteClient(discoveryServiceClient, Constants.Service.APP_FABRIC_HTTP,
+                                         HttpRequestConfig.DEFAULT,
+                                         Constants.Gateway.INTERNAL_API_VERSION_3);
+    this.retryStrategy = RetryStrategies.fromConfiguration(cConf, Constants.Service.TASK_WORKER + ".");
+    this.basePath = basePath;
+  }
+
+  /**
+   * Gets the location on the local filesystem for the given artifact. This method handles fetching the artifact as well
+   * as caching it.
+   *
+   * @param artifactId The ArtifactId of the artifact to fetch
+   * @return The Local Location for this artifact
+   * @throws ArtifactNotFoundException if the given artifact does not exist
+   * @throws IOException if there was an exception while fetching or caching the artifact
+   * @throws Exception if there was an unexpected error
+   */
+  public File getArtifact(ArtifactId artifactId) throws Exception {
+    LOG.debug("Fetching artifact info for {}", artifactId);
+    Long lastModifiedTimestamp = getCurrentLastModifiedTimestamp(artifactId);
+
+    File newJarLocation = Retries
+      .callWithRetries(() -> fetchArtifact(artifactId, lastModifiedTimestamp), retryStrategy);
+
+    // If the lastModifiedTimestamp is not null then we might need to delete the old cache.
+    if (lastModifiedTimestamp != null) {
+      // TODO (CDAP-18051): Migrate this cleanup task to its own service
+      deleteOldCache(artifactId, lastModifiedTimestamp, newJarLocation);
+    }
+    return newJarLocation;
+  }
+
+  /**
+   * Gets the location on the local filesystem for the directory that contains the unpacked artifact. This method
+   * handles fetching, caching and unpacking the artifact.
+   *
+   * @param artifactId The ArtifactId of the artifact to fetch and unpack
+   * @return The Local Location of the directory that contains the unpacked artifact files
+   * @throws ArtifactNotFoundException if the given artifact does not exist
+   * @throws IOException if there was an exception while fetching, caching or unpacking the artifact
+   * @throws Exception if there was an unexpected error
+   */
+  public File getAndUnpackArtifact(ArtifactId artifactId) throws Exception {
+    LOG.debug("Unpacking artifact {}", artifactId);
+    File jarLocation = getArtifact(artifactId);
+    File unpackDir = getUnpackLocalPath(artifactId, Long.valueOf(jarLocation.getName().split("\\.")[0]));
+    LOG.debug("Got unpack directory as {}", unpackDir);
+    if (!unpackDir.exists()) {
+      LOG.debug("Unpack directory doesn't exist, creating it now");
```

Comment:
Log with contextual info (e.g. which directory?) Also, you are unpacking a jar into the given directory instead of just creating the directory.

Reply:
done